vllm/engine/multiprocessing/mp_llm_engine.py

```diff
@@ -0,0 +1,259 @@
+
+        # Kick off core processing loop.
+        self.run_engine_loop()
+
+    @contextmanager
+    def make_data_socket(
+            self) -> Iterator[zmq.Socket]:  # type: ignore[name-defined]
+        socket = self.ctx.socket(zmq.constants.ROUTER)
+        try:
+            socket.bind(self.data_ipc_path)
+            yield socket
+        finally:
+            socket.close(linger=0)
+
+    def run_startup_loop(self) -> None:
+        """Loop over startup RPCStatupRequest from RPCClient."""
+
+        with self.make_data_socket() as socket:
+
+            # Loop until the RPCClient has all the data it needs.
+            client_is_ready = False
+            while not client_is_ready:
+                try:
+                    identity, message = socket.recv_multipart(copy=False)
+                    request: RPCStartupRequest = pickle.loads(message.buffer)
+
+                    # Handle the query from the Client.
+                    if request == RPCStartupRequest.GET_MODEL_CONFIG:
+                        response = self.engine.get_model_config()
+                    elif request == RPCStartupRequest.GET_DECODING_CONFIG:
+                        response = self.engine.get_decoding_config()
+                    elif request == RPCStartupRequest.GET_LORA_CONFIG:
+                        response = self.engine.get_lora_config()
+                    elif request == RPCStartupRequest.GET_SCHEDULER_CONFIG:
+                        response = self.engine.get_scheduler_config()
+                    elif request == RPCStartupRequest.GET_PARALLEL_CONFIG:
+                        response = self.engine.get_parallel_config()
+                    elif request == RPCStartupRequest.GET_TRACING_ENABLED:
+                        response = self.engine.is_tracing_enabled()
+                    elif request == RPCStartupRequest.IS_SERVER_READY:
+                        response = VLLM_RPC_SUCCESS_STR
+                    elif request == RPCStartupRequest.CLIENT_IS_READY:
+                        response = VLLM_RPC_SUCCESS_STR
+                        # Breakout of loop once client is ready.
+                        client_is_ready = True
+
+                    socket.send_multipart((identity, pickle.dumps(response)),
+                                          copy=False)
+
+                except Exception as e:
+                    socket.send_multipart((identity, pickle.dumps(e)),
+                                          copy=False)
+
+    def run_engine_loop(self) -> None:
+        while True:
+            # Block until there is a new request.
+            if not self.engine.has_unfinished_requests():
+                self.wait_for_new_input()
+
+            # Handle any new input from the input socket.
+            self.maybe_handle_new_input()
+
+            # Engine step.
+            request_outputs = self.engine.step()
+
+            # Stream results to output socket.
+            self.stream_outputs(request_outputs)
```

Comment:
to handle error propogatation, maybe we should add a `try-except` here, and send the exception back?